brokers/theia-plugin-broker/broker/broker.go

```diff
@@ -0,0 +1,224 @@
+	err = files.Unzip(archivePath, unpackedPath)
+	if err != nil {
+		return err
+	}
+
+	pj, err := broker.getPackageJSON(unpackedPath)
+	if err != nil {
+		return err
+	}
+
+	pluginImage, err := broker.getPluginImage(pj)
+	if err != nil {
+		return err
+	}
+	if pluginImage == "" {
+		// regular plugin
+		return broker.injectTheiaFile(meta, archivePath)
+	}
+	// remote plugin
+	return broker.injectTheiaRemotePlugin(meta, unpackedPath, pluginImage, pj)
+}
+
+func (broker *TheiaPluginBroker) getPackageJSON(pluginFolder string) (*packageJson, error) {
+	packageJSONPath := filepath.Join(pluginFolder, "package.json")
+	broker.PrintDebug("Reading package.json of Theia plugin from '%s'", packageJSONPath)
+	f, err := ioutil.ReadFile(packageJSONPath)
+	if err != nil {
+		return nil, err
+	}
+	pj := &packageJson{}
+	err = json.Unmarshal(f, pj)
+	return pj, err
+}
+
+func (broker *TheiaPluginBroker) getPluginImage(pj *packageJson) (string, error) {
+	if pj.Engines.CheRuntimeContainer != "" {
+		return pj.Engines.CheRuntimeContainer, nil
+	}
+	return "", nil
+}
+
+func (broker *TheiaPluginBroker) injectTheiaFile(meta model.PluginMeta, archivePath string) error {
+	broker.PrintDebug("Copying Theia plugin '%s:%s'", meta.ID, meta.Version)
+	pluginPath := filepath.Join("/plugins", fmt.Sprintf("%s.%s.theia", meta.ID, meta.Version))
+	err := files.CopyFile(archivePath, pluginPath)
+	if err != nil {
+		return err
+	}
+	tooling := &model.ToolingConf{}
+	return storage.AddPlugin(&meta, tooling)
+}
+
+func (broker *TheiaPluginBroker) injectTheiaRemotePlugin(meta model.PluginMeta, archiveFolder string, image string, pj *packageJson) error {
+	pluginFolderPath := filepath.Join("/plugins", fmt.Sprintf("%s.%s", meta.ID, meta.Version))
+	broker.PrintDebug("Copying Theia remote plugin '%s:%s' from '%s' to '%s'", meta.ID, meta.Version, archiveFolder, pluginFolderPath)
+	err := files.CopyResource(archiveFolder, pluginFolderPath)
+	if err != nil {
+		return err
+	}
+	tooling := &model.ToolingConf{
+		Containers: []model.Container{*containerConfig(image)},
+	}
+	broker.addPortToTooling(tooling, pj)
+	return storage.AddPlugin(&meta, tooling)
+}
+
+func containerConfig(image string) *model.Container {
+	c := model.Container{
+		Name:  "theiaPluginSidecar" + randomNumberAsString(),
+		Image: image,
+		Volumes: []model.Volume{
+			{
+				Name:      "projects",
+				MountPath: "/projects",
+			},
+			{
+				Name:      "plugins",
+				MountPath: "/plugins",
+			},
+		},
+	}
+	return &c
+}
+
+func (broker *TheiaPluginBroker) addPortToTooling(toolingConf *model.ToolingConf, pj *packageJson) {
+	port := findPort()
+	sPort := strconv.Itoa(port)
+	endpointName := "port" + sPort
+	var re = regexp.MustCompile(`[^a-z_0-9]+`)
+	prettyID := re.ReplaceAllString(pj.Publisher+"_"+pj.Name, `_`)
+	theiaEnvVar1 := "THEIA_PLUGIN_REMOTE_ENDPOINT_" + prettyID
+	theiaEnvVarValue := "ws://" + endpointName + ":" + sPort
```

Comment:
Will it work on completely https site like OSIO?

Reply:
yes